File: 20-valid-parentheses/valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        brack={')':'(','}':'{',']':'['}
        hash=set(['{','(','['])
        for i in s:
            if i in hash:
                stack.append(i)
            elif stack and brack[i]==stack[-1]:
                stack.pop()
            else:
                return False
        return len(stack)==0

File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        bool res = false;
        stack<char> stk;
        unordered_map<char, char> brackets ={{')','('}, {']','['}, {'}','{'}};
        unordered_set<char> openingBrackets = {'(', '{', '['}; 
        for(char c: s){
            if (openingBrackets.count(c)){
                stk.push(c);
            }
            else if(!stk.empty() && stk.top()==brackets[c]){
                stk.pop();
            }
            else{
                return false;
            }
        }
        return stk.size()==0;
    }
};